src/raftstore/store/fsm/store.rs

```diff
@@ -70,50 +85,494 @@ const PENDING_VOTES_CAP: usize = 20;
-    config.timer_tick_ms(cfg.raft_base_tick_interval.as_millis() / MIO_TICK_RATIO);
-    config.notify_capacity(cfg.notify_capacity);
-    config.messages_per_tick(cfg.messages_per_tick);
-    let event_loop = EventLoop::configured(config)?;
-    Ok(event_loop)
+pub struct RaftPoller<T: 'static, C: 'static> {
+    tag: String,
+    store_msg_buf: Vec<StoreMsg>,
+    peer_msg_buf: Vec<PeerMsg>,
+    previous_metrics: RaftMetrics,
+    timer: SlowTimer,
+    poll_ctx: PollContext<T, C>,
+    pending_proposals: Vec<RegionProposal>,
+    messages_per_tick: usize,
+}
+
+impl<T: Transport, C: PdClient> RaftPoller<T, C> {
+    fn handle_raft_ready(&mut self, peers: &mut [Box<PeerFsm>]) {
+        if !self.pending_proposals.is_empty() {
+            for prop in self.pending_proposals.drain(..) {
+                self.poll_ctx
+                    .apply_router
+                    .schedule_task(prop.region_id, ApplyTask::Proposal(prop));
+            }
+        }
+        if self.poll_ctx.need_flush_trans {
+            self.poll_ctx.trans.flush();
+            self.poll_ctx.need_flush_trans = false;
+        }
+        let ready_cnt = self.poll_ctx.ready_res.len();
+        self.poll_ctx.raft_metrics.ready.has_ready_region += ready_cnt as u64;
+        fail_point!("raft_before_save");
+        if !self.poll_ctx.kv_wb.is_empty() {
+            let mut write_opts = WriteOptions::new();
+            write_opts.set_sync(true);
+            let kv_wb = mem::replace(&mut self.poll_ctx.kv_wb, WriteBatch::new());
+            self.poll_ctx
+                .engines
+                .kv
+                .write_opt(kv_wb, &write_opts)
+                .unwrap_or_else(|e| {
+                    panic!("{} failed to save append state result: {:?}", self.tag, e);
+                });
+        }
+        fail_point!("raft_between_save");
+        if !self.poll_ctx.raft_wb.is_empty() {
+            let mut write_opts = WriteOptions::new();
+            write_opts.set_sync(self.poll_ctx.cfg.sync_log || self.poll_ctx.sync_log);
+            let raft_wb = mem::replace(
+                &mut self.poll_ctx.raft_wb,
+                WriteBatch::with_capacity(4 * 1024),
+            );
+            self.poll_ctx
+                .engines
+                .raft
+                .write_opt(raft_wb, &write_opts)
+                .unwrap_or_else(|e| {
+                    panic!("{} failed to save raft append result: {:?}", self.tag, e);
+                });
+        }
+        fail_point!("raft_after_save");
+        if ready_cnt != 0 {
+            let mut apply_tasks = Vec::with_capacity(ready_cnt);
+            let mut batch_pos = 0;
+            let mut ready_res = mem::replace(&mut self.poll_ctx.ready_res, Vec::default());
+            for (mut ready, invoke_ctx) in ready_res.drain(..) {
+                let region_id = invoke_ctx.region_id;
+                if peers[batch_pos].region_id() == region_id {
```

Comment:
empty if ???

Reply:
It's for the most path. After `likely/unlikely` are introduced, this can be rewritten.